### ISSUE TITLE
RouteParams parse errors

### ISSUE THREAD

**Issue by jangxyz** [redwoodjs/redwood#2868]
The RouteParams generic type in `web-routerRoutes.d.ts` fails to parse types correctly if there are untyped param together with typed ones.

Say I have a long pathname with multiple typed/untyped params:

`"/s/{bId:Int}/s/d/{year:Int}/{month:Int}/{monthDates}/e/{eId:Int}/s"`
as in
```typescript
const Routes = () => {
  return (
    <Router>
      <Route path="/home" page={HomePage} name="home" />
      <Route path="/s/{bId:Int}/s/d/{year:Int}/{month:Int}/{monthDates}/e/{eId:Int}/s"
        page={BadParamsPage} name="badParams" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

```

Note there is a `monthDates` param which does not have a specific parameter type. I am trying to denote multiple dates, as in `/2021/6/21,22,29,30/` so I am declaring it as string here.

However the generated type for `badParams` route method raises an error if I try to pass corresponding params object:

![Screen Shot 2021-06-22 at 12 58 17 AM](https://user-images.githubusercontent.com/52015/122792335-09e8bc00-d2f5-11eb-9d64-a30c2f9cdd17.png)

The generated type parses `monthDates` should be `{ "monthDates": string; }`, but instead is `{ "monthDates}/e/{eId": number; }`.

Here is the generated type file  (adding line breaks for readability):

```typescript
// web-routerRoutes.d.ts
import '@redwoodjs/router'

type ParamType<constraint> = constraint extends "Int" ? number : constraint extends "Boolean" ? boolean : constraint extends "Float" ? number : string;

// RouteParams
type RouteParams<Route> = 
  Route extends `${string}/{${infer Param}:${infer Constraint}}/${infer Rest}` // (1)
    ? { [Entry in Param]: ParamType<Constraint> } & RouteParams<`/${Rest}`>
    : Route extends `${string}/{${infer Param}:${infer Constraint}}` // (2)
      ? { [Entry in Param]: ParamType<Constraint> } 
      : Route extends `${string}/{${infer Param}}/${infer Rest}` // (3)
        ? { [Entry in Param]: string } & RouteParams<`/${Rest}`> 
        : {}

type QueryParams = Record<string | number, string | number | boolean>

declare module '@redwoodjs/router' {
  interface AvailableRoutes {
    // Only "<Route />" components with a "name" and "path" prop will be populated here.
    badParams: (params?: RouteParams<"/s/{bId:Int}/s/d/{year:Int}/{month:Int}/{monthDates}/e/{eId:Int}/s"> & QueryParams) => "/s/{bId:Int}/s/d/{year:Int}/{month:Int}/{monthDates}/e/{eId:Int}/s"
    home: (params?: RouteParams<"/home"> & QueryParams) => "/home"
  }
}
```

I am suspecting the `RouteParams`  type here.
Not totally sure how type inference in TypeScript works, but it seems like a fairly common error you see with regular expressions -- greedy matching.
RouteParams seems to digest the pathname from the beginning, and recursively re-infer the `Rest` part.

I can see it has generated types upto

```typescript
{ [Entry in "bId"]: number } & 
  { [Entry in "year"]: number } & 
    { [Entry in "month"]: number } & 
      RouteParams<"/{monthDates}/e/{eId:Int}/s">
```
correctly, but fails on matching the last part. It matches to condition (1), whereas it should match with (3). 

I think this should be fixed so that it correctly generates the type.

**Comment by thedavidprice** [comment]
Thanks for reporting this @jangxyz 

@dac09 @peterp suggestions about prioritization? My interpretation is this should be fixed for v1-rc milestone.

**Comment by Tobbe** [comment]
@jangxyz Are you on RW 0.34? Our latest release got a lot of typescript love, so if you're not on the latest release it's totally worth upgrading first 🙂

**Comment by jangxyz** [comment]
@Tobbe Yes this is done with a fresh 0.34.1 install.

**Comment by Tobbe** [comment]
@jangxyz

**Comment by keithtelliott** [comment]
I observed the same issue on 0.33, then upgraded today to 0.34 and still observe it. 

I am happy to jump in and start working toward a fix.   I’ve studied the code and am seeing what @jangxyz noted - very thorough notes @jangxyz - I appreciate it!   I have nothing more to add, other than an independent confirmation.

**Comment by dac09** [comment]
I've looked through this, and while its definitely an issue that should be solved - temporary work around is to add param types for all your route params. 

@keithtelliott very happy for you to take a shot - I know I wrote this, but I find it magical too, but I promise stare at it long enough it'll make sense.

Some test cases for you where I think its incorrect:
```
type mixed = RouteParams<"/a/{b:Int}/{c}/{d}">
type oneEach = RouteParams<"/a/{b:Int}/{c}/{d:Boolean}">
type oneEachDiffOrder = RouteParams<"/a/{b}/{c:Int}/{d:Boolean}">
type noTypesatall3 = RouteParams<"/a/{b}/{c}/{d}">
```

**Comment by thedavidprice** [comment]
@keithtelliott It think this would be a great PR effort for you to try! My 2 cents worth of advice is to start with a bare-minimum draft PR, which you can use to continue the conversation with @dac09 and @jangxyz 

The scope is nice and narrow. And you have all kinds of great support ready and waiting.

Go for it 🚀 (and let me know when/how I can help).

Sound good? 😉

**Comment by dac09** [comment]
I actually couldn't let go of this - put up a PR, which I believe fixes it. Totally would appreciate another pair of eyes. This time left comments - the template file does cause issues so whenever I use backticks, I had to wrap in `${"some_string_with_backticks"}` in template.js to make it work.

**Comment by thedavidprice** [comment]
Hi all! The PR has been merged and will be included in the v0.35 release going out by end of today. Could you all try it out and report back here as to whether or not things are resolved? 

Thanks!

**Comment by keithtelliott** [comment]
@dac09 and @thedavidprice I appreciate the encouragement!  I crawled into a learn-typescript cave and missed the opportunity to fix this bug (pardon my communication delay;  way to knock it out Danny!)

Nonetheless, I inspected the new code.  It is pleasing and satisfies my soul :)   Nice inline comments, in particular.  

Next, I will upgrade my app to v0.35 and functional test.  I’ll report back within a few days.
